openwisp_controller/config/static/config/js/widget.js

```diff
@@ -149,6 +166,11 @@
 
         // update raw value before form submit
         form.submit(function (e) {
+            // only submit form if the editor is clear of all validation errors
+            if (getEditorErrors(editor).length) {
+                e.preventDefault();
+                alert('Please correct all validation errors below');
```

Comment:
use `gettext` if present, this makes the code internationalization ready when we'll add this, do it as it's done here: https://github.com/openwisp/openwisp-controller/blob/master/openwisp_controller/config/static/config/js/preview.js#L55

Please read https://docs.djangoproject.com/en/3.1/topics/i18n/translation/#internationalization-in-javascript-code.